sign/cosi/cosi.go

```diff
@@ -1,452 +1,382 @@
-	// github.com/bford/golang-x-crypto/ed25519/cosi code.
-	*mask
-	// the message being co-signed
-	message []byte
-	// V_hat is the aggregated commit (our own + the children's)
-	aggregateCommitment kyber.Point
-	// challenge holds the challenge for this round
-	challenge kyber.Scalar
-
-	// the longterm private key CoSi will use during the response phase.
-	// The private key must have its public version in the list of publics keys
-	// given to CoSi.
-	private kyber.Scalar
-	// random is our own secret that we wish to commit during the commitment phase.
-	random kyber.Scalar
-	// commitment is our own commitment
-	commitment kyber.Point
-	// response is our own computed response
-	response kyber.Scalar
-	// aggregateResponses is the aggregated response from the children + our own
-	aggregateResponse kyber.Scalar
-}
-
-// NewCosi returns a new Cosi struct given the group, the longterm secret, and
-// the list of public keys. If some signers were not to be participating, you
-// have to set the mask using `SetMask` method. By default, all participants are
-// designated as participating. If you wish to specify which co-signers are
-// participating, use NewCosiWithMask
-func NewCosi(group kyber.Group, private kyber.Scalar, publics []kyber.Point) *CoSi {
-	cosi := &CoSi{
-		group:   group,
-		private: private,
-	}
-	// Start with an all-disabled participation mask, then set it correctly
-	cosi.mask = newMask(group, publics)
-	return cosi
-}
-
-// CreateCommitment creates the commitment of a random secret generated from the
-// given s stream. It returns the message to pass up in the tree. This is
-// typically called by the leaves.
-func (c *CoSi) CreateCommitment(s cipher.Stream) kyber.Point {
-	c.genCommit(s)
-	return c.commitment
+// Commit returns a random scalar v, generated from the given cipher stream,
+// and a corresponding commitment V = [v]G. If the given cipher stream is nil,
+// a random stream is used.
+func Commit(suite Suite, s cipher.Stream) (kyber.Scalar, kyber.Point) {
+	if s == nil {
+		s = random.Stream
+	}
+	random := suite.Scalar().Pick(s)
+	commitment := suite.Point().Mul(random, nil)
+	return random, commitment
 }
 
-// Commit creates the commitment / secret as in CreateCommitment and it also
-// aggregate children commitments from the children's messages.
-func (c *CoSi) Commit(s cipher.Stream, subComms []kyber.Point) kyber.Point {
-	// generate our own commit
-	c.genCommit(s)
-
-	// add our own commitment to the aggregate commitment
-	c.aggregateCommitment = c.group.Point().Add(c.group.Point().Null(), c.commitment)
-	// take the children commitments
-	for _, com := range subComms {
-		c.aggregateCommitment.Add(c.aggregateCommitment, com)
+// AggregateCommitments returns the sum of the given commitments and the
+// bitwise OR of the corresponding masks.
+func AggregateCommitments(suite Suite, commitments []kyber.Point, masks [][]byte) (kyber.Point, []byte, error) {
```

Comment:
So how do I get a `Mask` from this slice of byte ? It would be nice to see the flow from here to the signing which requires a  `*Mask`

Reply:
You can look at the test to see how it's achieved. After you have (received/computed) the aggregated mask `aggMask` you use `SetMask(aggMask)` to configure the `Mask` struct correctly. We assume that every node has already initialized a `Mask` struct in the beginning of the CoSi protocol run by default.